python/dolfinx/fem/nonlinearproblem.py

```diff
@@ -0,0 +1,122 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2021 Jørgen S. Dokken
+#
+# This file is part of DOLFINX (https://www.fenicsproject.org)
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+
+from typing import List
+import ufl
+from petsc4py import PETSc
+from dolfinx import fem, cpp
+import mpi4py
+
+
+class NonlinearProblem:
+    """Nonlinear problem class for solving the non-linear problem
+    F(u, v) = 0 for all v in V
+
+    """
+
+    def __init__(self, F: ufl.form.Form, u: fem.Function, bcs: List[fem.DirichletBC] = [],
+                 form_compiler_parameters={}, jit_parameters={}):
+        """Initialize class that sets up structures for solving the non-linear problem using Newton's method,
+        dF/du(u) du = -F(u)
+
+        Parameters
+        ----------
+        F
+            The PDE residual F(u, v)
+        u
+            The unknown
+        bcs
+            List of Dirichlet boundary conditions
+
+        form_compiler_parameters
+            Parameters used in FFCX compilation of this form. Run `ffcx --help` at
+            the commandline to see all available options. Takes priority over all
+            other parameter values, except for `scalar_type` which is determined by
+            DOLFINX.
+
+        jit_parameters
+            Parameters used in CFFI JIT compilation of C code generated by FFCX.
+            See `python/dolfinx/jit.py` for all available parameters.
+            Takes priority over all other parameter values.
+
+        .. code-block:: python
+            problem = LinearProblem(F, u, [bc0, bc1])
+        """
+        V = u.function_space
+        du = ufl.TrialFunction(V)
+        self.L = fem.form.Form(F, form_compiler_parameters=form_compiler_parameters, jit_parameters=jit_parameters)
+        # Create the Jacobian matrix, dF/du
+        self.a = fem.form.Form(ufl.derivative(F, u, du), form_compiler_parameters=form_compiler_parameters,
+                               jit_parameters=jit_parameters)
+        self.bcs = bcs
+
+        # Create matrix and vector to be used for assembly
+        # of the non-linear problem
+        self.matrix = fem.create_matrix(self.a)
```

Comment:
Not sure if this fits here well. I would say to either store the jacobian and residual state in the NonlinearProblem internally, and then do not ask in F() nor J() to have vector/matrix provided, but assemble into with internal state mat/vec.

Or maybe even better, do not store this state in the class, but cache it in the NewtonSolver. `NonlinerProblem` -- if you want to make it independent on the actual solver, should just hold residual and jacobian callbacks.

Reply:
> Not sure if this fits here well. I would say to either store the jacobian and residual state in the NonlinearProblem internally, and then do not ask in F() nor J() to have vector/matrix provided, but assemble into with internal state mat/vec.
> 
> Or maybe even better, do not store this state in the class, but cache it in the NewtonSolver. `NonlinerProblem` -- if you want to make it independent on the actual solver, should just hold residual and jacobian callbacks.

For this we would need to change the underlying C++ code. What do you think about this @garth-wells ?

We should probably also think about the design with respect to the SNESSolver (https://github.com/FEniCS/dolfinx/blob/4cdc251c325adbf884150aa00914da5eec1c2f40/python/test/unit/nls/test_newton.py#L190-L192) which uses the same design as our current NewtonSolver.

Reply:
If SNESSolver uses the same design as NewtonSolver that is good.

I prefer more how `NonlinearPDE_SNESProblem` looks. I.e. it computes jacobian form and residual form (if not provided explicitly), and user then calls `create_matrix` on this form fetched from the `NonlinearProblem`. It is just less error-prone, to keep less variables as the object state.

You can make NonlinearProblem look similar to  `NonlinearPDE_SNESProblem` and then both, our NewtonSolver and SNESSolver should be easily usable, no?

Reply:
I've now moved the Matrix and Vector construction into the python initializer of NewtonSolver. If the user wants to create this matrix themselves, they can always use the `cpp.nls.NewtonSolver`.

I figured I'd make a pull request with a SNESProblem after this is merged.